src/volatile_memory.rs

```diff
@@ -949,27 +1060,36 @@ where
     /// Does a volatile write of the value `v` to the address of this ref.
     #[inline(always)]
     pub fn store(&self, v: T) {
-        // SAFETY: Safe because we checked the address and size when creating this VolatileRef.
-        unsafe { write_volatile(self.addr, Packed::<T>(v)) };
-        self.bitmap.mark_dirty(0, size_of::<T>())
+        self.ptr_guard_mut(|ptr| {
+            // SAFETY: Safe because we checked the address and size when creating this VolatileRef.
+            unsafe { write_volatile(ptr as *mut Packed<T>, Packed::<T>(v)) };
+            self.bitmap.mark_dirty(0, self.len())
+        })
```

Comment:
Nice!

Reply:
@vireshk asked me to test this with virtiofsd again and I fear that this won't work nicely. One use case where virtiofsd needs the raw pointer is in order to issue `writev`/`readv` syscalls. These essentially receive a list of pointers + lens and will do scattered/gather read/writes. So given a list of slices, there needs to be a way to turn it into a list of pointers. With the previously sketched guard API that was doable by just keeping a list of guards. Now, that no longer works without nesting calls very deeply (and the code probably would not be pretty).

I understand that this was done because you had some concerns about the old "guard" not really guarding something. However, I am not sure how much safer the new proposal is. One could still, for example, accidentally access outside of the allowed length (or call .ptr_guard twice, nested and create a duplicated map). Do you have an example where you think the FnOnce solution is a better compared to the guard? Personally, I mostly understand it such that both cases still have unsafe inside of them but the FnOnce just makes the unsafe verification _slightly_ easier?

Of course, we could provide extra API to provide (and map) a list of pointers. But then we are introducing further new API variants at the cost of the just slightly easier unsafety verification. Also, where would we introduce that API on?

FWIW: The virtiofsd use-case could _maybe_ also be solvable by just providing a Read/Write impl somewhere that supports the Rust vectored read/write API. _Maybe_ something like https://github.com/rust-vmm/vm-virtio/pull/33 could be combined with the vectored read/write variants of `Read`/`Write` to solve virtiofsd needs. But the `writev`/`readv` examples are probably good hints that there are similar scenarios hidden somewhere that would require multiple slices to be mapped at the same time.

Reply:
Right, I see. I was mainly suggesting this because it looked simpler than the guard approach, but I agree that unless vm-memory provides better read(v)/write(v) abstractions this would make that quite complicated (fwiw I do have some sketches for something like that, but that's not useful at this time ofc). I kinda wanted to avoid the need for a `MappedVolatileSlice` thingy but I suppose there's no way around here.

Reply:
Okay, now that everyone agrees, I have moved back to the earlier version in somewhat better shape with less code. Hope it looks fine now.